Refactor template pod annotations

maiqueb opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:
This PR attempts to increase the readability of the pod annotation generation.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

maiqueb commented:
/cc @EdDev @phoracek

maiqueb commented on pkg/virt-controller/services/template.go:
```diff
@@ -1231,6 +1205,56 @@ func (t *templateService) renderLaunchManifest(vmi *v1.VirtualMachineInstance, t
+	annotationsList := map[string]string{
+		v1.DomainAnnotation: vmi.GetObjectMeta().GetName(),
+	}
+	for k, v := range copyVMIAnnotations(vmi) {
```
I'm unhappy w/ the `copyVMIAnnotations` name.

maiqueb replied:
> I'm unhappy w/ the `copyVMIAnnotations` name - it communicates intent, but is not in line with what the function does.

EdDev commented on pkg/virt-controller/services/template.go:
```diff
@@ -1547,44 +1546,64 @@ func getIfaceByName(vmi *v1.VirtualMachineInstance, name string) *v1.Interface {
+func getMultusCNIAnnotation(vmi *v1.VirtualMachineInstance) (string, error) {
+	var ifaceListMap []map[string]string
+
+	relevantNetworks := filterMultusNonDefaultNetworks(vmi)
```
How about "networks" or "multusNetworks" ?

maiqueb replied:
network can be pretty much anything, and multusNetworks can be multus non-default networks + multus default network. Both are wrong.

Renamed it to what I think is explicit enough.

EdDev commented on pkg/virt-controller/services/template.go:
```diff
@@ -1547,44 +1546,64 @@ func getIfaceByName(vmi *v1.VirtualMachineInstance, name string) *v1.Interface {
-	ifaceListMap := make([]map[string]string, 0)
-	cniAnnotations = make(map[string]string, 0)
+func getMultusCNIAnnotation(vmi *v1.VirtualMachineInstance) (string, error) {
+	var ifaceListMap []map[string]string
```
This name tell me little. Except that this is somehow related to interfaces, I am clueless without digging deep in what is put in it later on.
TBD

maiqueb replied:
Is it better now ?

EdDev commented on pkg/virt-controller/services/template.go:
```diff
@@ -1547,44 +1546,64 @@ func getIfaceByName(vmi *v1.VirtualMachineInstance, name string) *v1.Interface {
+}
 
-	next_idx := 0
+func mapifyMultusInterface(multusInterface *v1.Interface, networkName string, namespace string, interfaceSuffix int) map[string]string {
```
Sorry, but I lost track here.
- I do not understand the name of the function.
- Why there is no representation through a struct that is annotated. Having a proper name with ref to who/what defined it.
- Why do we need both `Interface` and `networkName`? The first already includes that interface/network name.

EdDev commented on pkg/virt-controller/services/template.go:
```diff
@@ -1547,44 +1546,64 @@ func getIfaceByName(vmi *v1.VirtualMachineInstance, name string) *v1.Interface {
+		"interface": fmt.Sprintf("net%d", interfaceSuffix+1),
+	}
+
+	if multusInterface != nil && multusInterface.MacAddress != "" {
```
In what case the interface can be nil? How can this be valid?

maiqueb replied:
We never check what we get out of the `getIfaceByName`. 

Just keeping the old behavior.

EdDev commented on pkg/virt-controller/services/template.go:
```diff
@@ -1547,44 +1546,64 @@ func getIfaceByName(vmi *v1.VirtualMachineInstance, name string) *v1.Interface {
+		// De-facto Standard doesn't define exact string format for
+		// MAC addresses pasted down to CNI.  Here we just pass through
+		// whatever the value our API layer accepted as legit.
+		// Note: while standard allows for 20-byte InfiniBand addresses,
+		// we forbid them in API.
```
How is this comment relevant here?

EdDev commented on pkg/virt-controller/services/template.go:
```diff
@@ -1547,44 +1546,64 @@ func getIfaceByName(vmi *v1.VirtualMachineInstance, name string) *v1.Interface {
+	return "", nil
+}
+
+func serializeMultusInterfaces(ifaceListMap []map[string]string) (string, error) {
```
I do not understand what this function adds here.
It saves no lines per what I see (the caller still needs to handle the error).

maiqueb replied:
OK.

EdDev reviewed (changes requested):
I think the annotations merit their own file now. And extract the tests to test them in particular.

I would even go beyond and claim that each annotation group should have its own file. You can even create an annotation package here.

maiqueb commented:
Addressed @EdDev 's comments in [this forced push](https://github.com/kubevirt/kubevirt/compare/8e8ce1ce691af7169727d9b888d3589ab2b5fc0f..03247612e222ee5875620210dc849f897ec38c62).

Rebased latest master [this forced push](https://github.com/kubevirt/kubevirt/compare/03247612e222ee5875620210dc849f897ec38c62..8d1b97ee6a5652fefea643b567d60046efb7c7fe).

EdDev commented on pkg/virt-controller/services/multus_annotations.go:
```diff
@@ -0,0 +1,81 @@
+	}
+}
+
+func filterMultusNonDefaultNetworks(vmi *v1.VirtualMachineInstance) []v1.Network {
```
Can you pass in only the networks?

EdDev commented on pkg/virt-controller/services/multus_annotations.go:
```diff
@@ -0,0 +1,81 @@
+}
+
+func newMultusAnnotationStruct(vmi *v1.VirtualMachineInstance, network v1.Network, podInterfaceName string) multusNetworkAnnotation {
+	multusIface := getIfaceByName(vmi, network.Name)
```
I think you can move here `getIfaceByName` as well.

maiqueb replied:
Done.

EdDev reviewed (commented):
Could you please explain in the commit message shorty what you do? (e.g. extract to functions, move to a new file, etc...)

maiqueb reviewed (commented):
Thanks for the thorough review. I appreciate the time you're spending, @EdDev .

EdDev commented on pkg/virt-controller/services/multus_annotations.go:
```diff
@@ -0,0 +1,90 @@
+func getIfaceByName(vmi *v1.VirtualMachineInstance, name string) *v1.Interface {
+	for _, iface := range vmi.Spec.Domain.Devices.Interfaces {
+		if iface.Name == name {
+			return &iface
```
It is unsafe to return a reference to a loop iterator value variable. See [here](https://github.com/golang/go/wiki/CommonMistakes#using-reference-to-loop-iterator-variable).

It has been treated by #4751

maiqueb replied:
I'm just moving that code from A to B, but OK. Addressed it in a new commit on top of the others.

EdDev commented on pkg/virt-controller/services/multus_annotations.go:
```diff
@@ -0,0 +1,90 @@
+
+	multusNonDefaultNetworks := filterMultusNonDefaultNetworks(vmi.Spec.Networks)
+	for i, network := range multusNonDefaultNetworks {
+		serializedInterface := newMultusAnnotationData(vmi, network, fmt.Sprintf("net%d", i+1))
```
Just wandering why you call a return struct value "serialized". I think that "serialized" refers to something that can be send over to a remote destination through a protocol, like a string, json, xml. Am I getting confused here?

If I am not mistaken, the serialization occurs when `json.Marshal` is called, right?

maiqueb replied:
This was addressed already in https://github.com/kubevirt/kubevirt/compare/8d1b97ee6a5652fefea643b567d60046efb7c7fe..c1709d426e838e63a95baffc30a5a3b6a907f610

EdDev replied:
The link does not work, can you update it?

maiqueb replied:
Works for me. Nothing I can do I'm afraid.

But, at the end of the day, it's been addressed; resolving.

EdDev commented on pkg/virt-controller/services/multus_annotations.go:
```diff
@@ -0,0 +1,90 @@
+	if len(multusNetworkAnnotationPool) > 0 {
+		multusNetworksAnnotation, err := json.Marshal(multusNetworkAnnotationPool)
+		if err != nil {
+			return "", fmt.Errorf("failed to create JSON list from CNI interface map %v", multusNetworkAnnotationPool)
```
nit: CNI, map ==> multus, pool

maiqueb replied:
Done.

EdDev commented on pkg/virt-controller/services/multus_annotations.go:
```diff
@@ -33,21 +33,37 @@ type multusNetworkAnnotation struct {
+	pool []multusNetworkAnnotation
+}
+
+func (mnap *multusNetworkAnnotationPool) add(multusNetworkAnnotation multusNetworkAnnotation) {
```
nit: `mnap` is hard to pronounce (which is funny that I even try when reading). I would just use `p` but this is just an opinionated nit, feel free to ignore.

EdDev commented on pkg/virt-controller/services/multus_annotations.go:
```diff
@@ -33,21 +33,37 @@ type multusNetworkAnnotation struct {
+	return len(mnap.pool) == 0
+}
+
+func (mnap multusNetworkAnnotationPool) toString() (string, error) {
```
Ahh, if you could have used `String`, it will fit the `Stringer` interface, which `fmt` can use then. But that error is in the way.

maiqueb replied:
I could return an empty string instead of an error if we really want to implement the `Stringer` interface. 

I don't think it's worth it, but I can compromise.

EdDev replied:
No need from my part, I just shared my thought.

EdDev reviewed (commented):
Looks very nice!

Just a few inline comments.

EdDev reviewed (approved):
Thank you!

maiqueb commented:
/retest

maiqueb commented:
/cc @enp0s3 could you take a look at this PR ?

enp0s3 commented:
@maiqueb Sorry for the late response, will review it

enp0s3 reviewed (approved):
Thanks you!
/approve

enp0s3 commented:
/hold

enp0s3 commented:
@maiqueb Would you prefer to add unit tests for the multus annotation in this PR or in a separate PR?

maiqueb commented:
> @maiqueb Would you prefer to add unit tests for the multus annotation in this PR or in a separate PR?

I can do it now. I'll remove the hold; please lgtm if you're on board with them @enp0s3 .

Thanks for the review.

/hold cancel

EdDev commented on pkg/virt-controller/services/multus_annotations_test.go:
```diff
@@ -0,0 +1,100 @@
+ *
+ */
+
+package services
```
Testing private functions is coupling the tests to the internal implementation and that may be worse than having none.
I'm unclear about what can be done in this specific case, I would just try to examine if these tests can be checked from the formal API (public functions).

maiqueb replied:
As I see it, 2 choices here: either get this in as is (with these unit tests, that check the un-exported functions), or figure that out in a follow up PR. 

Multus annotation generation is already tested by the template generation, in https://github.com/kubevirt/kubevirt/blob/master/pkg/virt-controller/services/template_test.go#L715

EdDev replied:
So are those tests cover the scenarios you presented here?

maiqueb replied:
More than that.

EdDev replied:
Too short answer form me to understand.
If these scenarios are already covered, I do not understand why you created a duplication now.

Ideally, if you can check these scenarios through the exported functions, that is preferred. If that means to move the existing tests that are relevant to here, great.
I just ask you to check the options of this, if you see it is too "heavy" then we can defer it to later.

maiqueb replied:
Let's defer to later then.

EdDev commented on pkg/virt-controller/services/multus_annotations_test.go:
```diff
@@ -0,0 +1,100 @@
+			}
+		})
+
+		It("is not empty", func() {
+			Expect(multusAnnotationPool.isEmpty()).To(BeFalse())
+		})
```
I think that the check should show that a pool creation is successful. The empty check can be part of it.

maiqueb replied:
What would you test there ? Pool creation gives you an empty pool: which does not fit this context.

It's unclear to me what you want.

EdDev replied:
I'm trying to say that the check that it is empty is odd as a test.
Creating an empty pool and checking it is empty, sounds good.
Creating a non empty pool and checking it is not empty, sounds good.

If you really think this makes sense, I won't insist.

maiqueb replied:
> "
> Creating an empty pool and checking it is empty, sounds good.

I'm doing this.

 I'm severely confused. In the line above you say -and I quote: "I'm trying to say that the check that it is empty is odd as a test." and now you say creating an empty pool and checking it is empty is good. You can't go left and right at the same time; which one is it ?

> Creating a non empty pool and checking it is not empty, sounds good.

I'm doing this.

> 
> If you really think this makes sense, I won't insist.

OK. Let's leave it at that then.

maiqueb commented on pkg/virt-controller/services/multus_annotations_test.go:
```diff
@@ -0,0 +1,100 @@
+
+	Context("a multus annotation pool with no elements", func() {
+		BeforeEach(func() {
+			multusAnnotationPool = multusNetworkAnnotationPool{}
```
Is it resetting the variable between tests ?

EdDev replied:
So why set it outside at all?

maiqueb replied:
I don't follow. Please help me understand.

EdDev replied:
Define it in the test body as you do nothing specific with it in the fixture.

maiqueb replied:
What is the advantage ? The disadvantage being that I need to repeat myself in each test case.

maiqueb replied:
Furthermore, I think this is how BDD (which ginkgo implements) works; these tests are in the scope of an 'empty' pool. The context's setup should create that empty pool.

maiqueb commented on pkg/virt-controller/services/multus_annotations_test.go:
```diff
@@ -0,0 +1,100 @@
+	Context("a multus annotation pool with no elements", func() {
+		BeforeEach(func() {
+			multusAnnotationPool = multusNetworkAnnotationPool{}
+			// for checking the multus annotations, all we need is the namespace
+			vmi = v1.VirtualMachineInstance{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "testvmi", Namespace: "default", UID: "1234",
+				},
+			}
+			network = v1.Network{
+				NetworkSource: v1.NetworkSource{
+					Multus: &v1.MultusNetwork{NetworkName: "test1"},
+				},
+			}
```
It can.

maiqueb replied:
Done.

maiqueb commented on pkg/virt-controller/services/multus_annotations_test.go:
```diff
@@ -0,0 +1,100 @@
+		It("generate a null string", func() {
+			multusAnnotation, err := multusAnnotationPool.toString()
+			Expect(multusAnnotation).To(BeIdenticalTo("null"))
+			Expect(err).NotTo(HaveOccurred())
```
Done.

maiqueb commented on pkg/virt-controller/services/multus_annotations_test.go:
```diff
@@ -0,0 +1,100 @@
+			expectedString := `[{"interface":"net1","name":"test1","namespace":"namespace1"}]`
+			multusAnnotation, err := multusAnnotationPool.toString()
+			Expect(multusAnnotation).To(BeIdenticalTo(expectedString))
+			Expect(err).NotTo(HaveOccurred())
```
Done.

maiqueb commented:
@EdDev is there an action point from me here ? Would you rather have me remove these unit tests ? I think they add some value.

@enp0s3 are you OK with the proposed unit tests ?

EdDev commented:
> @EdDev is there an action point from me here ? Would you rather have me remove these unit tests ? I think they add some value.

I consider testing private functions a bad practice which costs more pain than gain.
It makes sense to accept this when no other choices exists, however, in this case you are saying that the same code flow is already examined by other tests (which I hope exercise the public/exported functions).
IMO you need to reason for make the test code so coupled with the implementation and that is the text that is missing from my part.

I have not voted here as I do not want to block this change, it is much better than before. I have even lgtm it when these tests were not included.

maiqueb commented:
> > @EdDev is there an action point from me here ? Would you rather have me remove these unit tests ? I think they add some value.
> 
> I consider testing private functions a bad practice which costs more pain than gain.
> It makes sense to accept this when no other choices exists, however, in this case you are saying that the same code flow is already examined by other tests (which I hope exercise the public/exported functions).
> IMO you need to reason for make the test code so coupled with the implementation and that is the text that is missing from my part.

Is it enough If I export the methods on the pool thing ? 

What reasoning are you after ? This multus annotation pool is, literally a one trick pony. It only needs to generate a string. And I don't imagine it being called from anywhere else.

> 
> I have not voted here as I do not want to block this change, it is much better than before. I have even lgtm it when these tests were not included.

enp0s3 reviewed (approved):
Thank you!
/approve
